File: Helper/Bridge/Library.cs
#pragma warning disable 0,
// ReSharper disable

using Helper;
using Helper.Utility;

namespace Helper.Bridge {

	public unsafe class Library {

		#region structor

		private IntPtr? mHandle = null;

		private SymbolTable? mSymbol = null;

		// ----------------

		public Library (
		) {
		}

		#endregion

		#region state

		public Boolean State (
		) {
			return this.mHandle is not null;
		}

		public void Open (
			String path
		) {
			GF.AssertTest(!this.State());
			var handle = ExternalLibrary.Kernel32.LoadLibrary($"{path}.");
			var symbol = new SymbolTable() {
				execute = null!,
			};
			try {
				symbol.execute = Marshal.GetDelegateForFunctionPointer<Interface.execute>(ExternalLibrary.Kernel32.GetProcAddress(handle, SymbolNameTable.execute));
			}
			catch (Exception) {
				ExternalLibrary.Kernel32.FreeLibrary(handle);
				throw;
			}
			this.mHandle = handle;
			this.mSymbol = symbol;
			return;
		}

		public void Close (
		) {
			GF.AssertTest(this.State());
			var result = ExternalLibrary.Kernel32.FreeLibrary(this.mHandle.AsNotNull());
			GF.AssertTest(result);
			this.mHandle = null;
			this.mSymbol = null;
			return;
		}

		#endregion

		#region interface

		public Interface.String* Execute (
			Interface.Callback**   callback,
			Interface.String**     script,
			Interface.StringList** argument,
			Interface.String**     result
		) {
			GF.AssertTest(this.State());
			return this.mSymbol.AsNotNull().execute(callback, script, argument, result);
		}

		#endregion

	}

}

File: ShellGUI/lib/bridge/library.dart
import '/common.dart';
import '/bridge/interface.dart' as Interface; // ignore: library_prefixes
import '/bridge/symbol.dart';
import 'dart:io';
import 'dart:ffi' as ffi;

// ----------------

class Library {

  ffi.DynamicLibrary? _handle = null;
  SymbolTable?        _symbol = null;

  // ----------------

  Boolean state(
  ) {
    return this._handle != null;
  }

  Void open(
    String path,
  ) {
    assertAlways(!this.state());
    var pathAbsolute = path;
    if (Platform.isWindows) {
      pathAbsolute += '.';
    }
    var handle = ffi.DynamicLibrary.open(pathAbsolute);
    var symbol = SymbolTable();
    try {
      symbol.execute = handle.lookup<ffi.NativeFunction<Interface.execute>>(SymbolNameTable.execute);
    }
    catch (e) {
      handle.close();
      rethrow;
    }
    this._handle = handle;
    this._symbol = symbol;
    return;
  }

  Void close(
  ) {
    assertAlways(this.state());
    this._handle!.close();
    this._handle = null;
    this._symbol = null;
    return;
  }

  // ----------------

  ffi.Pointer<Interface.String> execute(
    ffi.Pointer<ffi.Pointer<Interface.Callback>>   callback,
    ffi.Pointer<ffi.Pointer<Interface.String>>     script,
    ffi.Pointer<ffi.Pointer<Interface.StringList>> argument,
    ffi.Pointer<ffi.Pointer<Interface.String>>     result,
  ) {
    assertAlways(this.state());
    return this._symbol!.execute.asFunction<Interface.execute>()(callback, script, argument, result);
  }

}
